### PR TITLE
feat: Add off-screen indicator to WorldWaypointDistanceFeature

### PR DESCRIPTION
**Pull request by HalcyonSuoh**

### CODE REVIEW

File: common/src/main/java/com/wynntils/wynn/model/CompassModel.java [hunk 1]
```diff
@@ -57,18 +59,28 @@ public static Optional<WaypointPoi> getCompassWaypoint() {
         return Optional.empty();
     }
 
+    public static Texture getTargetIcon() {
+        return targetIcon;
+    }
+
     public static void setDynamicCompassLocation(Supplier<Location> compassSupplier) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/wynn/model/CompassModel.java:66]: Why? Imo, we should still take a texture as argument. Ofc, you can have a method without texture, which defaults to Texture.WAYPOINT. 

We don't need a texture supplier here as even `DynamicIconPoi` has a static icon texture.

File: common/src/main/java/com/wynntils/gui/screens/maps/MainMapScreen.java [hunk 2]
```diff
@@ -233,7 +234,12 @@ public boolean mouseClicked(double mouseX, double mouseY, int button) {
             if (hovered != null && !(hovered instanceof TerritoryPoi)) {
                 McUtils.playSound(SoundEvents.EXPERIENCE_ORB_PICKUP);
                 if (hovered.hasStaticLocation()) {
-                    CompassModel.setCompassLocation(new Location(hovered.getLocation()));
+                    if (hovered instanceof IconPoi) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/gui/screens/maps/MainMapScreen.java]:
```suggestion
                    if (hovered instanceof IconPoi iconPoi) {
```

This way, you don't need to cast again.

File: common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java [hunk 3]
```diff
@@ -82,61 +96,170 @@ public void onRenderLevelPost(RenderLevelEvent.Post event) {
...
+            centerRelativePosition = new Vec3(minY / m, minY, 0);
+        }
+
+        // bring the position back to normal screen space (top left origin point)
+        return new Vec2(
+                (float) (centerRelativePosition.x + centerPoint.x), (float) (centerRelativePosition.y + centerPoint.y));
+    }
+
+    private boolean isInBound(Vec3 position, Window window) {
+        return 0 < position.x
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java]:
```suggestion
        return position.x > 0
```

File: common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java [hunk 4]
```diff
@@ -82,61 +96,170 @@ public void onRenderLevelPost(RenderLevelEvent.Post event) {
...
+                && position.x < window.getGuiScaledWidth()
+                && 0 < position.y
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java]:
```suggestion
                && position.y > 0
```

File: common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java (hunk 3 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java]: This way, position is always on the left side. This improves readability.

File: common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java [hunk 5]
```diff
@@ -82,61 +96,170 @@ public void onRenderLevelPost(RenderLevelEvent.Post event) {
...
-                        VerticalAlignment.Middle,
-                        textShadow);
-    }
+        float displayPositionX;
+        float displayPositionY;
+
+        Vec2 intersectPoint = getBoundingIntersectPoint(screenCoord, event.getWindow());
+        Texture icon = CompassModel.getTargetIcon();
+
+        if (intersectPoint == null) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java:118]:
```suggestion
       // The set waypoint is visible on the screen, so we render the icon + distance
        if (intersectPoint == null) {
```

File: common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java [hunk 6]
```diff
@@ -82,61 +96,170 @@ public void onRenderLevelPost(RenderLevelEvent.Post event) {
...
+
+        Window window = McUtils.window();
+
+        return new Vec3(
+                (float) ((ndc.x + 1.0f) / 2.0f) * window.getGuiScaledWidth(),
+                (float) ((1.0f - ndc.y) / 2.0f) * window.getGuiScaledHeight(),
+                (float) ndc.z);
+    }
+
+    // tldr: draw a line from screen center to the target's screenspace coordinate
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java]:
```suggestion
    // draw a line from screen center to the target's screenspace coordinate
```

File: common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java [hunk 7]
```diff
@@ -82,61 +96,170 @@ public void onRenderLevelPost(RenderLevelEvent.Post event) {
...
+        // (x2 - x1) is the equivalent of the x position
+        // c is the line's y-intercept, but line pass through origin, so this will be 0
+        // finalize to y = mx, or x = y/m
+
+        if (centerRelativePosition.x > 0) {
+            centerRelativePosition = new Vec3(maxX, maxX * m, 0);
+        } else {
+            centerRelativePosition = new Vec3(minX, minX * m, 0);
+        }
+        if (centerRelativePosition.y > maxY) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java:249]:
```suggestion

        if (centerRelativePosition.y > maxY) {
```

**Review by kristofbolyai — COMMENTED** [review]
I have some nit-picks to improve code quality, but this looks great!

**Review by kristofbolyai — APPROVED** [review]
Looks good to me! Thank you, this looks really great in-game! 

Also, you mentioned in discord that math could be optimized. It sure could be, but this whole feature does not even show up on CPU/Memory profiles :)

File: common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java [hunk 8]
```diff
@@ -82,61 +96,172 @@ public void onRenderLevelPost(RenderLevelEvent.Post event) {
...
+        // finalize to y = mx, or x = y/m
+
+        if (centerRelativePosition.x > 0) {
+            centerRelativePosition = new Vec3(maxX, maxX * m, 0);
+        } else {
+            centerRelativePosition = new Vec3(minX, minX * m, 0);
+        }
+
+        if (centerRelativePosition.y > maxY) {
+            centerRelativePosition = new Vec3(maxY / m, maxY, 0);
```
Comment (magicus) [inline common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java:250]: Are we 100% sure we can't get a division by zero here? If the angle is 0 then m will be 0. But I must admit I got a bit lost on the maths here.

Comment (HalcyonSuoh) [reply common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java:250]: > Are we 100% sure we can't get a division by zero here? If the angle is 0 then m will be 0. But I must admit I got a bit lost on the maths here.

centerRelativePosition.y > maxY is the case where the line intersect with bottom bounding, in which the only case where slope is zero is probably where resolution's height being 0, I don't think such edge case exist in any playable situation

Comment (HalcyonSuoh) [reply common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java:250]: on second thought, situation might be possible if bottom bounding goes higher than screen center, so it think it's intersecting with right and bottom bounding simultaneously, I'll fix it if I can get a crash out of it

**Pull request merged**